contracts/wallet/WalletFactory.sol

```diff
@@ -56,38 +69,135 @@ contract WalletFactory is Owned, Managed {
-        } else {
-            // initialise the wallet with the owner and the modules
-            BaseWallet(wallet).init(_owner, _modules);
+        configureWallet(BaseWallet(wallet), _owner, _modules, _label, address(0));
+    }
+
+    /**
+     * @dev Lets the manager create a wallet for an owner account.
+     * The wallet is initialised with a list of modules, a first guardian, and an ENS..
+     * The wallet is created using the CREATE opcode.
+     * @param _owner The account address.
+     * @param _modules The list of modules.
+     * @param _label ENS label of the new wallet, e.g. franck.
+     * @param _guardian The guardian address.
+     */
+    function createWalletWithGuardian(
+        address _owner,
+        address[] calldata _modules,
+        string calldata _label,
+        address _guardian
+    )
+        external
+        onlyManager
+        guardianStorageDefined
+    {
+        validateInputs(_owner, _modules, _label);
+        Proxy proxy = new Proxy(walletImplementation);
+        address payable wallet = address(proxy);
+        configureWallet(BaseWallet(wallet), _owner, _modules, _label, _guardian);
+    }
+
+    /**
+     * @dev Lets the manager create a wallet for an owner account at a specific address.
+     * The wallet is initialised with a list of modules and an ENS.
+     * The wallet is created using the CREATE2 opcode.
+     * @param _owner The account address.
+     * @param _modules The list of modules.
+     * @param _label ENS label of the new wallet, e.g. franck.
+     * @param _salt The salt.
+     */
+    function createCounterfactualWallet(
+        address _owner,
+        address[] calldata _modules,
+        string calldata _label,
+        bytes32 _salt
+    )
+        external
+        onlyManager
+    {
+        validateInputs(_owner, _modules, _label);
+        bytes32 newsalt = keccak256(abi.encodePacked(_salt, _owner, _modules));
+        bytes memory code = abi.encodePacked(type(Proxy).creationCode, uint256(walletImplementation));
+        address payable wallet;
+        // solium-disable-next-line security/no-inline-assembly
+        assembly {
+            wallet := create2(0, add(code, 0x20), mload(code), newsalt)
+            if iszero(extcodesize(wallet)) { revert(0, returndatasize) }
         }
-        emit WalletCreated(wallet, _owner);
+        configureWallet(BaseWallet(wallet), _owner, _modules, _label, address(0));
+    }
+
+    /**
+     * @dev Lets the manager create a wallet for an owner account at a specific address.
+     * The wallet is initialised with a list of modules, a first guardian, and an ENS.
+     * The wallet is created using the CREATE2 opcode.
+     * @param _owner The account address.
+     * @param _modules The list of modules.
+     * @param _label ENS label of the new wallet, e.g. franck.
+     * @param _guardian The guardian address.
+     * @param _salt The salt.
+     */
+    function createCounterfactualWalletWithGuardian(
```

Comment:
Just a final thing on overloads, I think it makes sense for the following to be overloads to at most 2 functions:
`createWallet(address _owner, address[] calldata _modules, string calldata _label)`
`function createWalletWithGuardian(address _owner, address[] calldata _modules, string calldata _label, address _guardian)`
`function createCounterfactualWallet(address _owner, address[] calldata _modules, string calldata _label, bytes32 _salt)`
`createCounterfactualWalletWithGuardian(address _owner, address[] calldata _modules, string calldata _label,         address _guardian, bytes32 _salt)`

Reply:
We've experienced many problems with overloading and the different clients used in our stack in the past (Web3J, web3.swift, web3.js and ethers.js). Maybe this has been fixed but I'd prefer to avoid overloading as much as I can.